os/drivers/lcd/st7789.c

```diff
@@ -0,0 +1,901 @@
+
+static inline void lcd_unlock(FAR struct spi_dev_s *dev)
+{
+	(void)SPI_LOCK(dev, false);
+}
+
+/****************************************************************************
+ * Name: st7789_sendcmd
+ *
+ * Description:
+ *   Send a command to the driver.
+ *
+ ****************************************************************************/
+static inline void st7789_sendcmd(FAR struct st7789_dev_s *dev, uint8_t cmd)
+{
+#ifdef CONFIG_LCD_ST7789_3WIRE
+	uint16_t txbuf;
+	/* Add command prefix (9th bit shoudl be 0 ) */
+	txbuf = LCD_ST7789_CMD_PREFIX | cmd;
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	SPI_SEND(dev->spi, txbuf);
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, false);
+#else
+	uint8_t spi_cmd[4];
+	spi_cmd[0] = cmd;
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	dev->config->cmddata(0);
+	SPI_SNDBLOCK(dev->spi, spi_cmd, 1);
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, false);
+#endif
+}
+
+/****************************************************************************
+ * Name: st7789_senddata
+ *
+ * Description:
+ *   Send init data to the driver.
+ * data - tx buffer
+ * size - total bytes of words
+ *        word size(8 bits)
+ ****************************************************************************/
+
+static inline void st7789_senddata(FAR struct st7789_dev_s *dev, uint8_t *txbuf, int size)
+{
+#ifdef CONFIG_LCD_ST7789_3WIRE
+	uint16_t txbuf;
+	/* Add data prefix (9th bit shoudl be 1 ) */
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	for (int j = 0; j < size; j += 2) {
+		/* Take care of correct byte order. */
+
+		rowbuff[j] = LCD_ST7789_DATA_PREFIX | (uint16_t) txbuf[j + 1 + (i * size)];
+		rowbuff[j + 1] = LCD_ST7789_DATA_PREFIX | (uint16_t) txbuf[j + (i * size)];
+	}
+	SPI_SNDBLOCK(dev->spi, rowbuff, size);
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, false);
+#else
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	dev->config->cmddata(1);
+	SPI_SNDBLOCK(dev->spi, txbuf, size);
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, false);
+#endif
+}
+
+/****************************************************************************
+ * Name: st7789_getdata
+ *
+ * Description:
+ *   Get data from LCD.
+ *
+ * buf - rx buffer
+ * size - total bytes of words
+ *        word size(8 bits)
+ ****************************************************************************/
+
+static inline void st7789_getdata(FAR struct st7789_dev_s *dev, uint8_t *rxbuf, int size)
+{
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, true);
+	dev->config->cmddata(1);
+	SPI_RECVBLOCK(dev->spi, rxbuf, size);
+	SPI_SELECT(dev->spi, SPIDEV_DISPLAY, false);
+}
+
+/****************************************************************************
+ * Name: st7789_sleep
+ *
+ * Description:
+ *   Sleep or wake up the driver.
+ *
+ ****************************************************************************/
+
+static void st7789_sleep(FAR struct st7789_dev_s *dev, bool sleep)
+{
+	lcd_lock(dev->spi);
+	st7789_sendcmd(dev, sleep ? ST7789_SLPIN : ST7789_SLPOUT);
+	lcd_unlock(dev->spi);
+	up_mdelay(120);
```

Comment:
Is this required? If yes, do we need `up_mdelay` instead of `sleep`? Because `up_mdelay` is busy waiting, it's not good at performance.